test/ts/unit/collateralized_simple_interest_terms_contract.ts

```diff
@@ -0,0 +1,812 @@
+                    );
+                });
+            });
+        });
+    });
+
+    describe("#registerRepayment", () => {
+        describe("agent who is not RepaymentRouter calls registerRepayment", () => {
+            it("should throw", async () => {
+                await expect(
+                    termsContract.registerRepayment.sendTransactionAsync(
+                        ARBITRARY_AGREEMENT_ID,
+                        PAYER,
+                        BENEFICIARY,
+                        Units.ether(1),
+                        mockToken.address,
+                        { from: ATTACKER }, // the message is coming from an attacker.
+                    ),
+                ).to.eventually.be.rejectedWith(REVERT_ERROR);
+            });
+        });
+
+        describe("RepaymentRouter calls registerRepayment", () => {
+            before(async () => {
+                const principalTokenIndex = new BigNumber(0);
+                const principalAmount = Units.ether(1);
+                const interestRate = Units.percent(1);
+                const amortizationUnitType = new BigNumber(0);
+                const termLengthInAmortizationUnits = new BigNumber(3);
+
+                const termsContractParameters = SimpleInterestParameters.pack(
+                    principalTokenIndex,
+                    principalAmount,
+                    interestRate,
+                    amortizationUnitType,
+                    termLengthInAmortizationUnits,
+                );
+
+                await mockRegistry.mockGetBeneficiaryReturnValueFor.sendTransactionAsync(
+                    ARBITRARY_AGREEMENT_ID,
+                    BENEFICIARY,
+                );
+
+                await mockRegistry.mockGetTermsReturnValueFor.sendTransactionAsync(
+                    ARBITRARY_AGREEMENT_ID,
+                    termsContract.address,
+                    termsContractParameters,
+                );
+
+                await mockRegistry.mockGetTermsContractReturnValueFor.sendTransactionAsync(
+                    ARBITRARY_AGREEMENT_ID,
+                    termsContract.address,
+                );
+
+                await mockTokenRegistry.mockGetTokenAddressByIndex.sendTransactionAsync(
+                    new BigNumber(0),
+                    mockToken.address,
+                );
+            });
+
+            describe("...with a different `tokenAddress` than expected by the terms contract", () => {
+                let dummyToken: MockERC20TokenContract;
+
+                before(async () => {
+                    const mockTokenContractTruffle = await mockTokenContract.new();
+                    const mockTokenContractWeb3 = web3.eth
+                        .contract(mockTokenContractTruffle.abi)
+                        .at(mockTokenContractTruffle.address);
+                    dummyToken = new MockERC20TokenContract(mockTokenContractWeb3, TX_DEFAULTS);
+
+                    await router.repay.sendTransactionAsync(
```

Comment:
This test will pass, but for the wrong reasons.  The payer needs to grant the token transfer proxy an allowance of dummyToken, or else it won't ever reach `registerRepayment`

Reply:
I think this is handled on line 122:

```javascript
await mockToken.mockAllowanceFor.sendTransactionAsync(
    PAYER,
    mockTokenTransferProxy.address,
    Units.ether(3),
);
```

Reply:
👍